data_prep/generate_annual_averages.py

```diff
@@ -0,0 +1,191 @@
+
+log = logging.getLogger(__name__)
+logging.basicConfig(stream=sys.stdout, level=4)
+
+def iter_matching(dirpath, regexp):
+    # http://stackoverflow.com/questions/4639506/os-walk-with-regex
+    """Generator yielding all files under `dirpath` whose absolute path
+       matches the regular expression `regexp`.
+       Usage:
+
+           >>> for filename in iter_matching('/', r'/home.*\.bak'):
+           ....    # do something
+    """
+    for dir_, dirnames, filenames in os.walk(dirpath):
+        for filename in filenames:
+            abspath = os.path.join(dir_, filename)
+            if regexp.match(abspath):
+                yield abspath
+
+def create_annual_avg_file(in_fp, out_fp, variable):
+    '''
+    Generates file with annual averages for the given variable across
+    all time within an input file
+
+    Parameters:
+        in_fp: input file path
+        out_fp: output file path
+        variable (str): name of the variable which is being processed
+
+    '''
+    supported_vars = {
+        'cddETCCDI', 'csdiETCCDI', 'cwdETCCDI', 'dtrETCCDI', 'fdETCCDI',
+        'gslETCCDI', 'idETCCDI', 'prcptotETCCDI', 'r10mmETCCDI', 'r1mmETCCDI',
+        'r20mmETCCDI', 'r95pETCCDI', 'r99pETCCDI', 'rx1dayETCCDI',
+        'rx5dayETCCDI', 'sdiiETCCDI', 'suETCCDI', 'thresholds', 'tn10pETCCDI',
+        'tn90pETCCDI', 'tnnETCCDI', 'tnxETCCDI', 'trETCCDI', 'tx10pETCCDI',
+        'tx90pETCCDI', 'txnETCCDI', 'txxETCCDI', 'wsdiETCCDI', 'tasmin',
+        'tasmax', 'pr'
+    }
+
+    if variable not in supported_vars:
+        raise Exception("Unsupported variable: can't yet process {}".format(variable))
+
+    if not os.path.exists(os.path.dirname(out_fp)):
+        os.makedirs(os.path.dirname(out_fp))
+
+    cdo = Cdo()
+    cdo.yearmean(input=in_fp, output=out_fp)
+
+def update_annual_avg_file_global_metadata(nc):
+    '''
+    Modifies global metadata of generated yearmean NetCDF file
+    '''
+    title = nc.getncattr('title')
+    new_title = 'Annual Average of Daily ' + title
+    new_frequency = 'yr'
+    nc.setncatts({'title': new_title, 'frequency': new_frequency})
+
+def update_annual_avg_file_variable_metadata(nc, variable):
+    '''
+    Modifies metadata of one variable of generated yearmean NetCDF file
+    '''
+    long_name = nc.variables[variable].getncattr('long_name')
+    new_long_name = 'Annual Average ' + long_name
+    nc.variables[variable].setncattr('long_name', new_long_name)
+
+def has_valid_time_bounds(bounds_var):
+    '''
+    Verifies if the input NetCDF has a valid, and populated, time_bnds variable
+    '''
+    # bounds_var might be a np.masked_array, and if any elements are masked, it's bad
+    if (hasattr(bounds_var, 'mask') and (np.ma.is_masked(bounds_var[:]))):
+        return False
+    # or, bounds_var might be full of zeros, also bad
+    elif (not np.any(bounds_var)):
+        return False
+    else:
+        return True
+
+def update_annual_avg_file_time_metadata(nc, start_year=None):
```

Comment:
I feel like the time var and the time_bnds var should be split out into two separate functions here...